File: go/internal/presenter/presenter.go
// Package presenter converts orchestrator snapshots to HTTP API payloads and HTML.
package presenter

import (
	"encoding/json"
	"fmt"
	"html"
	"strings"
	"time"

	"symphony/internal/config"
	"symphony/internal/dashboard"
	"symphony/internal/orchestrator"
)

// SnapshotError represents a snapshot fetch error.
type SnapshotError int

const (
	SnapshotUnavailable SnapshotError = iota
	SnapshotTimeout
)

// StatePayload builds the JSON payload for GET /api/v1/state.
func StatePayload(snapshot *orchestrator.Snapshot, err SnapshotError) map[string]any {
	generatedAt := time.Now().UTC().Format(time.RFC3339)

	if snapshot == nil {
		code := "snapshot_unavailable"
		message := "Snapshot unavailable"
		if err == SnapshotTimeout {
			code = "snapshot_timeout"
			message = "Snapshot timed out"
		}
		return map[string]any{
			"generated_at": generatedAt,
			"error": map[string]any{
				"code":    code,
				"message": message,
			},
		}
	}

	running := make([]map[string]any, 0, len(snapshot.Running))
	for _, entry := range snapshot.Running {
		running = append(running, projectRunningEntry(&entry))
	}
	retrying := make([]map[string]any, 0, len(snapshot.Retrying))
	for _, entry := range snapshot.Retrying {
		retrying = append(retrying, projectRetryEntry(&entry))
	}

	return map[string]any{
		"generated_at": generatedAt,
		"counts": map[string]any{
			"running":  len(snapshot.Running),
			"retrying": len(snapshot.Retrying),
		},
		"running":      running,
		"retrying":     retrying,
		"codex_totals": snapshot.CodexTotals,
		"rate_limits":  json.RawMessage(snapshot.RateLimits),
	}
}

// IssuePayload builds the JSON payload for GET /api/v1/:identifier.
func IssuePayload(snapshot *orchestrator.Snapshot, issueIdentifier string, settings *config.Settings) map[string]any {
	var runningEntry *orchestrator.RunningSnapshot
	var retryEntry *orchestrator.RetrySnapshot

	for i := range snapshot.Running {
		if snapshot.Running[i].Identifier == issueIdentifier {
			runningEntry = &snapshot.Running[i]
			break
		}
	}
	for i := range snapshot.Retrying {
		if snapshot.Retrying[i].Identifier == issueIdentifier {
			retryEntry = &snapshot.Retrying[i]
			break
		}
	}

	if runningEntry == nil && retryEntry == nil {
		return nil
	}

	status := "retrying"
	if runningEntry != nil {
		status = "running"
	}

	issueID := ""
	if runningEntry != nil {
		issueID = runningEntry.IssueID
	} else if retryEntry != nil {
		issueID = retryEntry.IssueID
	}

	workspacePath := ""
	workerHost := ""
	if runningEntry != nil {
		workspacePath = runningEntry.WorkspacePath
		workerHost = runningEntry.WorkerHost
	} else if retryEntry != nil {
		workspacePath = retryEntry.WorkspacePath
		workerHost = retryEntry.WorkerHost
	}
	if workspacePath == "" && settings != nil {
		workspacePath = settings.Workspace.Root + "/" + issueIdentifier
	}
	if workspacePath == "" {
		workspacePath = issueIdentifier
	}

	restartCount := 0
	currentRetry := 0
	if retryEntry != nil {
		if retryEntry.Attempt > 1 {
			restartCount = retryEntry.Attempt - 1
		}
		currentRetry = retryEntry.Attempt
	}

	var lastError *string
	if retryEntry != nil && retryEntry.Error != "" {
		lastError = &retryEntry.Error
	}

	result := map[string]any{
		"issue_identifier": issueIdentifier,
		"issue_id":         issueID,
		"status":           status,
		"workspace": map[string]any{
			"path": workspacePath,
			"host": workerHost,
		},
		"attempts": map[string]any{
			"restart_count":        restartCount,
			"current_retry_attempt": currentRetry,
		},
		"logs":          map[string]any{"codex_session_logs": []any{}},
		"recent_events": []any{},
		"last_error":    lastError,
		"tracked":       map[string]any{},
	}

	if runningEntry != nil {
		result["running"] = projectRunningEntry(runningEntry)
	}
	if retryEntry != nil {
		result["retry"] = projectRetryEntry(retryEntry)
	}

	return result
}

// RenderDashboardHTML renders the full HTML dashboard page.
func RenderDashboardHTML(snapshot *orchestrator.Snapshot, settings *config.Settings) string {
	terminalView := dashboard.FormatSnapshot(snapshot, settings, 0, 115)

	var b strings.Builder
	b.WriteString("<!DOCTYPE html><html lang=\"en\"><head><meta charset=\"utf-8\">")
	b.WriteString("<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">")
	b.WriteString("<title>Symphony Observability</title>")
	b.WriteString("<link rel=\"stylesheet\" href=\"/dashboard.css\">")
	b.WriteString("</head><body>")
	b.WriteString("<main class=\"app-shell\"><section class=\"dashboard-shell\">")

	// Hero
	b.WriteString("<header class=\"hero-card\"><div class=\"hero-grid\"><div>")
	b.WriteString("<p class=\"eyebrow\">Symphony Observability</p>")
	b.WriteString("<h1 class=\"hero-title\">Operations Dashboard</h1>")
	b.WriteString("<p class=\"hero-copy\">Current state, retry pressure, token usage, and orchestration health.</p>")
	b.WriteString("</div></div></header>")

	// Metrics
	b.WriteString("<section class=\"metric-grid\">")
	b.WriteString(fmt.Sprintf("<article class=\"metric-card\"><p class=\"metric-label\">Running</p><p class=\"metric-value numeric\">%d</p></article>", len(snapshot.Running)))
	b.WriteString(fmt.Sprintf("<article class=\"metric-card\"><p class=\"metric-label\">Retrying</p><p class=\"metric-value numeric\">%d</p></article>", len(snapshot.Retrying)))
	b.WriteString(fmt.Sprintf("<article class=\"metric-card\"><p class=\"metric-label\">Total tokens</p><p class=\"metric-value numeric\">%s</p></article>", formatInt(snapshot.CodexTotals.TotalTokens)))
	b.WriteString("</section>")

	// Terminal status
	b.WriteString("<section class=\"section-card\"><div class=\"section-header\"><div><h2 class=\"section-title\">Terminal status</h2></div></div>")
	b.WriteString("<div class=\"terminal-frame\"><pre class=\"terminal-dashboard\">")
	b.WriteString(html.EscapeString(terminalView))
	b.WriteString("</pre></div></section>")

	// Running table
	b.WriteString("<section class=\"section-card\"><div class=\"section-header\"><div><h2 class=\"section-title\">Running sessions</h2></div></div>")
	b.WriteString("<div class=\"table-wrap\"><table class=\"data-table\">")
	b.WriteString("<thead><tr><th>Issue</th><th>State</th><th>Session</th><th>Tokens</th><th>Runtime</th></tr></thead><tbody>")
	if len(snapshot.Running) == 0 {
		b.WriteString("<tr><td colspan=\"5\" class=\"muted\">No active agents</td></tr>")
	} else {
		for _, entry := range snapshot.Running {
			b.WriteString(fmt.Sprintf("<tr><td>%s</td><td>%s</td><td class=\"mono\">%s</td><td class=\"numeric\">%s</td><td class=\"numeric\">%ds</td></tr>",
				html.EscapeString(entry.Identifier),
				html.EscapeString(entry.State),
				html.EscapeString(entry.SessionID),
				formatInt(entry.CodexTotalTokens),
				entry.RuntimeSeconds))
		}
	}
	b.WriteString("</tbody></table></div></section>")

	// Retry table
	b.WriteString("<section class=\"section-card\"><div class=\"section-header\"><div><h2 class=\"section-title\">Retry queue</h2></div></div>")
	b.WriteString("<div class=\"table-wrap\"><table class=\"data-table\">")
	b.WriteString("<thead><tr><th>Issue</th><th>Attempt</th><th>Due in</th><th>Error</th></tr></thead><tbody>")
	if len(snapshot.Retrying) == 0 {
		b.WriteString("<tr><td colspan=\"4\" class=\"muted\">No queued retries</td></tr>")
	} else {
		for _, entry := range snapshot.Retrying {
			id := entry.Identifier
			if id == "" {
				id = entry.IssueID
			}
			b.WriteString(fmt.Sprintf("<tr><td>%s</td><td class=\"numeric\">%d</td><td class=\"numeric\">%dms</td><td>%s</td></tr>",
				html.EscapeString(id),
				entry.Attempt,
				entry.DueInMs,
				html.EscapeString(entry.Error)))
		}
	}
	b.WriteString("</tbody></table></div></section>")

	b.WriteString("</section></main></body></html>")
	return b.String()
}

func projectRunningEntry(entry *orchestrator.RunningSnapshot) map[string]any {
	return map[string]any{
		"issue_id":             entry.IssueID,
		"identifier":           entry.Identifier,
		"state":                entry.State,
		"worker_host":          entry.WorkerHost,
		"workspace_path":       entry.WorkspacePath,
		"session_id":           entry.SessionID,
		"codex_app_server_pid": entry.CodexAppServerPID,
		"codex_input_tokens":   entry.CodexInputTokens,
		"codex_output_tokens":  entry.CodexOutputTokens,
		"codex_total_tokens":   entry.CodexTotalTokens,
		"turn_count":           entry.TurnCount,
		"started_at":           entry.StartedAt.Format(time.RFC3339),
		"last_codex_timestamp": formatTimePtr(entry.LastCodexTimestamp),
		"last_codex_message":   entry.LastCodexMessage,
		"last_codex_event":     entry.LastCodexEvent,
		"runtime_seconds":      entry.RuntimeSeconds,
	}
}

func projectRetryEntry(entry *orchestrator.RetrySnapshot) map[string]any {
	return map[string]any{
		"issue_id":       entry.IssueID,
		"attempt":        entry.Attempt,
		"due_in_ms":      entry.DueInMs,
		"identifier":     entry.Identifier,
		"error":          entry.Error,
		"worker_host":    entry.WorkerHost,
		"workspace_path": entry.WorkspacePath,
	}
}

func formatTimePtr(t *time.Time) any {
	if t == nil {
		return nil
	}
	return t.Format(time.RFC3339)
}

func formatInt(n int64) string {
	if n >= 1_000_000 {
		return fmt.Sprintf("%.1fM", float64(n)/1_000_000)
	}
	if n >= 1_000 {
		return fmt.Sprintf("%.1fK", float64(n)/1_000)
	}
	return fmt.Sprintf("%d", n)
}

File: rust/src/presenter.rs
use chrono::{DateTime, TimeDelta, Utc};
use serde_json::{Value as JsonValue, json};

use crate::config::Settings;
use crate::orchestrator::{RetrySnapshot, RunningSnapshot, Snapshot};
use crate::status_dashboard;

pub fn state_payload(snapshot: Result<&Snapshot, SnapshotError>) -> JsonValue {
    let generated_at = generated_at();
    match snapshot {
        Ok(snapshot) => json!({
            "generated_at": generated_at,
            "counts": {
                "running": snapshot.running.len(),
                "retrying": snapshot.retrying.len()
            },
            "running": snapshot.running.iter().map(project_running_entry).collect::<Vec<_>>(),
            "retrying": snapshot.retrying.iter().map(project_retry_entry).collect::<Vec<_>>(),
            "codex_totals": snapshot.codex_totals,
            "rate_limits": snapshot.rate_limits
        }),
        Err(SnapshotError::Timeout) => json!({
            "generated_at": generated_at,
            "error": {
                "code": "snapshot_timeout",
                "message": "Snapshot timed out"
            }
        }),
        Err(SnapshotError::Unavailable) => json!({
            "generated_at": generated_at,
            "error": {
                "code": "snapshot_unavailable",
                "message": "Snapshot unavailable"
            }
        }),
    }
}

pub fn issue_payload(
    snapshot: &Snapshot,
    issue_identifier: &str,
    settings: Option<&Settings>,
) -> Option<JsonValue> {
    let running = snapshot
        .running
        .iter()
        .find(|entry| entry.identifier == issue_identifier);
    let retry = snapshot
        .retrying
        .iter()
        .find(|entry| entry.identifier.as_deref() == Some(issue_identifier));

    if running.is_none() && retry.is_none() {
        return None;
    }

    let workspace_path = running
        .and_then(|entry| entry.workspace_path.clone())
        .or_else(|| retry.and_then(|entry| entry.workspace_path.clone()))
        .or_else(|| {
            settings.map(|settings| {
                settings
                    .workspace
                    .root
                    .join(issue_identifier)
                    .to_string_lossy()
                    .to_string()
            })
        })
        .unwrap_or_else(|| issue_identifier.to_string());

    Some(json!({
        "issue_identifier": issue_identifier,
        "issue_id": running.map(|entry| entry.issue_id.clone()).or_else(|| retry.map(|entry| entry.issue_id.clone())),
        "status": if running.is_some() { "running" } else { "retrying" },
        "workspace": {
            "path": workspace_path,
            "host": running.and_then(|entry| entry.worker_host.clone()).or_else(|| retry.and_then(|entry| entry.worker_host.clone()))
        },
        "attempts": {
            "restart_count": retry.map(|entry| entry.attempt.saturating_sub(1)).unwrap_or(0),
            "current_retry_attempt": retry.map(|entry| entry.attempt).unwrap_or(0)
        },
        "running": running.map(project_running_issue_body),
        "retry": retry.map(project_retry_issue_body),
        "logs": { "codex_session_logs": [] },
        "recent_events": running.map(project_recent_events).unwrap_or_else(|| json!([])),
        "last_error": retry.and_then(|entry| entry.error.clone()),
        "tracked": {}
    }))
}

pub fn render_dashboard_html(snapshot: &Snapshot, settings: &Settings) -> String {
    let terminal_view = status_dashboard::format_snapshot_content_for_test(
        Some(snapshot),
        settings,
        0.0,
        Some(115),
    );
    format!(
        "<!DOCTYPE html><html><head><meta charset=\"utf-8\"><title>Symphony</title><link rel=\"stylesheet\" href=\"/dashboard.css\"><script defer src=\"/vendor/phoenix_html/phoenix_html.js\"></script><script defer src=\"/vendor/phoenix/phoenix.js\"></script><script defer src=\"/vendor/phoenix_live_view/phoenix_live_view.js\"></script></head><body><main data-phx-main><header><h1>Symphony</h1><p class=\"status-badge-live\">Live</p><p class=\"status-badge-offline\">Offline</p></header><section><pre class=\"terminal-dashboard\">{}</pre></section></main></body></html>",
        escape_html(&terminal_view)
    )
}

#[derive(Clone, Copy, Debug)]
pub enum SnapshotError {
    Timeout,
    Unavailable,
}

fn project_running_entry(entry: &RunningSnapshot) -> JsonValue {
    json!({
        "issue_id": entry.issue_id,
        "issue_identifier": entry.identifier,
        "state": entry.state,
        "worker_host": entry.worker_host,
        "workspace_path": entry.workspace_path,
        "session_id": entry.session_id,
        "turn_count": entry.turn_count,
        "last_event": entry.last_codex_event,
        "last_message": status_dashboard::humanize_codex_message(entry.last_codex_message.as_ref()),
        "started_at": iso8601(Some(entry.started_at)),
        "last_event_at": iso8601(entry.last_codex_timestamp),
        "tokens": {
            "input_tokens": entry.codex_input_tokens,
            "output_tokens": entry.codex_output_tokens,
            "total_tokens": entry.codex_total_tokens
        }
    })
}

fn project_retry_entry(entry: &RetrySnapshot) -> JsonValue {
    json!({
        "issue_id": entry.issue_id,
        "issue_identifier": entry.identifier,
        "attempt": entry.attempt,
        "due_at": due_at_iso8601(entry.due_in_ms),
        "error": entry.error,
        "worker_host": entry.worker_host,
        "workspace_path": entry.workspace_path
    })
}

fn project_running_issue_body(entry: &RunningSnapshot) -> JsonValue {
    json!({
        "worker_host": entry.worker_host,
        "workspace_path": entry.workspace_path,
        "session_id": entry.session_id,
        "turn_count": entry.turn_count,
        "state": entry.state,
        "started_at": iso8601(Some(entry.started_at)),
        "last_event": entry.last_codex_event,
        "last_message": status_dashboard::humanize_codex_message(entry.last_codex_message.as_ref()),
        "last_event_at": iso8601(entry.last_codex_timestamp),
        "tokens": {
            "input_tokens": entry.codex_input_tokens,
            "output_tokens": entry.codex_output_tokens,
            "total_tokens": entry.codex_total_tokens
        }
    })
}

fn project_retry_issue_body(entry: &RetrySnapshot) -> JsonValue {
    json!({
        "attempt": entry.attempt,
        "due_at": due_at_iso8601(entry.due_in_ms),
        "error": entry.error,
        "worker_host": entry.worker_host,
        "workspace_path": entry.workspace_path
    })
}

fn project_recent_events(entry: &RunningSnapshot) -> JsonValue {
    match entry.last_codex_timestamp {
        Some(timestamp) => json!([{
            "at": iso8601(Some(timestamp)),
            "event": entry.last_codex_event,
            "message": status_dashboard::humanize_codex_message(entry.last_codex_message.as_ref())
        }]),
        None => json!([]),
    }
}

fn due_at_iso8601(due_in_ms: u64) -> String {
    iso8601(Some(Utc::now() + TimeDelta::milliseconds(due_in_ms as i64))).unwrap()
}

fn iso8601(timestamp: Option<DateTime<Utc>>) -> Option<String> {
    timestamp.map(|value| value.format("%Y-%m-%dT%H:%M:%SZ").to_string())
}

fn generated_at() -> String {
    Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string()
}

fn escape_html(value: &str) -> String {
    value
        .replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::{CliOverrides, Settings};
    use crate::orchestrator::{PollingSnapshot, TokenTotals};
    use crate::workflow::LoadedWorkflow;

    fn settings() -> Settings {
        Settings::from_workflow(
            &LoadedWorkflow {
                config: serde_yaml::from_str("tracker:\n  kind: memory\n").unwrap(),
                prompt_template: String::new(),
                prompt: String::new(),
            },
            &CliOverrides::default(),
        )
        .unwrap()
    }

    fn snapshot() -> Snapshot {
        Snapshot {
            running: vec![RunningSnapshot {
                issue_id: "issue-http".to_string(),
                identifier: "MT-HTTP".to_string(),
                state: "In Progress".to_string(),
                worker_host: None,
                workspace_path: None,
                session_id: Some("thread-http".to_string()),
                codex_app_server_pid: None,
                codex_input_tokens: 4,
                codex_output_tokens: 8,
                codex_total_tokens: 12,
                turn_count: 7,
                started_at: Utc::now(),
                last_codex_timestamp: None,
                last_codex_message: Some(JsonValue::String("rendered".to_string())),
                last_codex_event: Some("notification".to_string()),
                runtime_seconds: 42,
            }],
            retrying: vec![RetrySnapshot {
                issue_id: "issue-retry".to_string(),
                attempt: 2,
                due_in_ms: 5000,
                identifier: Some("MT-RETRY".to_string()),
                error: Some("boom".to_string()),
                worker_host: None,
                workspace_path: None,
            }],
            codex_totals: TokenTotals {
                input_tokens: 4,
                output_tokens: 8,
                total_tokens: 12,
                seconds_running: 42,
            },
            rate_limits: Some(json!({"primary": {"remaining": 11}})),
            polling: PollingSnapshot {
                checking: false,
                next_poll_in_ms: Some(5000),
                poll_interval_ms: 30000,
            },
        }
    }

    #[test]
    fn state_and_issue_payloads_humanize_messages() {
        let snapshot = snapshot();
        let state = state_payload(Ok(&snapshot));
        assert_eq!(state["running"][0]["last_message"], "rendered");

        let issue = issue_payload(&snapshot, "MT-HTTP", Some(&settings())).unwrap();
        assert_eq!(issue["running"]["last_message"], "rendered");
    }

    #[test]
    fn dashboard_html_embeds_terminal_dashboard_and_assets() {
        let html = render_dashboard_html(&snapshot(), &settings());
        assert!(html.contains("/dashboard.css"));
        assert!(html.contains("/vendor/phoenix_html/phoenix_html.js"));
        assert!(html.contains("terminal-dashboard"));
        assert!(html.contains("SYMPHONY STATUS"));
    }
}
